Final Working Python Code Using AIF Extracted from Vessel Centerlines

Caffetaria opened the pull request.

tbirdso commented:
@Caffetaria Thank you for your hard work this summer and for putting this together!

As a quick heads up, I will be working with @aylward over the next couple of weeks towards making your recent work in this branch available as a Python `pip`-installable package.

I am going to refactor the commit history in `itk-version`. To explain the refactoring goals I have in mind:
- Avoid commits that introduce and then delete files. When we add a data file in one commit and then remove it in the next commit, such as in https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/commit/c905681e1f72b7f59a9bcc75c8ec9304386de9ca and https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/commit/1da70d92c60a2c29ccfd8fce367161e4cd3e9405, the file still exists in commit history and can make the repository size much larger. An easy solution here is to squash commits with `git reset --soft origin/main` so that those changes cancel out and are ultimately removed from history.
- Compress `NIFTI` files. ITK can read directly from compressed `.nii.gz` files, so we can compress the proposed `.nii` files to `.nii.gz` to save space.
- Reduce or avoid adding data files to the repository altogether to minimize the repository download size. We often host large 3D/4D data on external sites such as `data.kitware.com` and can direct users to download the data archive as part of setup in the `README` file.

I've backed up existing commit history in the branch https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/tree/itk-version-aria. Please feel free to comment here if you have questions or thoughts on changes that could be made. Thanks!

EDIT: I have finished refactoring. Please find data files at https://data.kitware.com/#folder/64da58971c6956f5031e433d going forward.

tbirdso committed:
ENH: Add `.gitignore`

Add `.gitignore` to ignore pycache files

tbirdso committed:
ENH: Update Python code to use AIF from Vessel Centerlines

Changes:
- Update scripts to work with Computed Tomography (CT) images
- Converted SimpleITK to ITK workflow
- Updated methods to use arterial input function (AIF) extracted from
  vessel centerlines

co-authored-by: Aria Vikram <ariavikram@gmail.com>

tbirdso committed:
ENH: Update README with project overview and instructions

Update README with project overview and instructions to run perfusion
imaging methods as a local application from the repository source.

co-authored-by: Aria Vikram <ariavikram@gmail.com>

tbirdso committed:
ENH: Maintain `Data` as an external resource

Changes:
- Ignores local `Data` directory in Git to avoid adding large data files
  to git history
- Updates README to reference upstream perfusion image data files
  provided for examples and testing

co-authored-by: Aria Vikram <ariavikram@gmail.com>

tbirdso commented on lib/ParamsCalculator/aif.py:
```diff
@@ -1,18 +1,495 @@
+from scipy.optimize import least_squares
+
+def DSC_aif(mask, conc, bolus, size, config, device, vessels):
+    nSlice = 69
```
@aylward @Caffetaria Could you comment on the motivation for using this hardcoded slice value?

My guess is that this always extracts time point 68 from the 4D image. But, our data only appears to have 16(?) time points, leading to an error here.

EDIT: See related failure in CI development in https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/tree/python-module: https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/actions/runs/5867124421/job/15907251087

```py
  def DSC_aif(mask, conc, bolus, size, config, device, vessels):
        nSlice = 69
>       conc_reshaped = conc[nSlice - 1, :, :, :]
E       IndexError: index 68 is out of bounds for axis 0 with size 18

src/perfusion_analysis_toolbox/lib/ParamsCalculator/aif.py:8: IndexError
```

aylward replied:
This is probably the z-slice from the 3D volumes that comprise the 4D images.   This number should come from an atlas-registration process, that she didn't have time to implement.  Perhaps keep it at 69 and add the comment that it is the z-slice that should intersect the major intercranial arterial vessels.   Ultimately there won't be a slice, but a volume-of-interest in 3D defined by the region of expected vessel locations transcribed from a vascular atlas.

tbirdso replied:
> This is probably the z-slice from the 3D volumes that comprise the 4D images. This number should come from an atlas-registration process, that she didn't have time to implement.

Makes sense, and that's fine. We can move forward with the value in this PR, but for subsequent testing I'm unclear on whether we have an input data set that works with this parameter.

If this is an arbitrary z-slice index, would it be equally valid to select a different z-slice index that fits with our test data so that the application does not fail at this point? I'm unfamiliar with the consequences of which z-slice we select here.

aylward replied:
We had two sets of testing data: high res and low res.

The high res 4D volume should be 16 time points with 376x376x170 volumes.   The nSlice value probably corresponds to the high res testing data

If you're using the low res 4D volume (size 16 time points with 138x138x18 volumes).  The nSlice value should be 7 for the low res testing data

Hope this helps.

tbirdso replied:
> We had two sets of testing data: high res and low res.

This is a piece I was missing. I see the [high-res mask](https://data.kitware.com/#item/64da58a81c6956f5031e433f) but I'm not seeing a corresponding input volume in our [sample data](https://data.kitware.com/#folder/64da58971c6956f5031e433d). The high-res mask has size 170x376x376, whereas `CTP04-4D_reg.nii.gz` has size 16x18x138x138.

Could you please point me towards the high-res 4D input volume?

aylward commented:
Uploading now - but it is 1.3G!

I suggest using the low res data for testing to save transfer and
computation time.

s

On Tue, Aug 15, 2023 at 10:45 AM Tom Birdsong ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In lib/ParamsCalculator/aif.py
> <https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/pull/2#discussion_r1294692201>
> :
>
> >  from builtins import object
> +from scipy import linalg
> +from scipy.optimize import least_squares
> +from scipy.signal import convolve
> +import matplotlib.pyplot as plt
> +from scipy.ndimage import label
> +from scipy.cluster.hierarchy import linkage, fcluster
> +from scipy.spatial.distance import pdist
> +from scipy.optimize import least_squares
> +
> +def DSC_aif(mask, conc, bolus, size, config, device, vessels):
> +    nSlice = 69
>
> We had two sets of testing data: high res and low res.
>
> This is a piece I was missing. I see the high-res mask
> <https://data.kitware.com/#item/64da58a81c6956f5031e433f> but I'm not
> seeing a corresponding input volume in our sample data
> <https://data.kitware.com/#folder/64da58971c6956f5031e433d>. The high-res
> mask has size 170x376x376, whereas CTP04-4D_reg.nii.gz has size
> 16x18x138x138.
>
> Could you please point me towards the high-res 4D input volume?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/pull/2#discussion_r1294692201>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEJL6SY4UMWKDFQF3RWVTXVODPDANCNFSM6AAAAAAZQM3LXM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> .com>
>


-- 
Stephen R. Aylward, Ph.D.
Chair, MONAI Advisory Board
Senior Director, Strategic Initiatives, Kitware

tbirdso commented:
> Uploading now - but it is 1.3G!

> I suggest using the low res data for testing to save transfer and
computation time.

@aylward Thanks!

Now that data sources are clear, I'm in favor of moving forward with merging the current PR and making any changes to parameters and methods in a subsequent PR. Does that work for you?

aylward commented:
+1!

...and the data is still uploading (am traveling today for an offsite
meeting).

On Tue, Aug 15, 2023 at 11:16 AM Tom Birdsong ***@***.***>
wrote:

> Uploading now - but it is 1.3G!
>
> I suggest using the low res data for testing to save transfer and
> computation time.
>
> @aylward <https://github.com/aylward> Thanks!
>
> Now that data sources are clear, I'm in favor of moving forward with
> merging the current PR and making any changes to parameters and methods in
> a subsequent PR. Does that work for you?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KitwareMedical/Perfusion-Analysis-Toolbox/pull/2#issuecomment-1679111419>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEJLZJMSMQWS52JWJUNE3XVOHEBANCNFSM6AAAAAAZQM3LXM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Stephen R. Aylward, Ph.D.
Chair, MONAI Advisory Board
Senior Director, Strategic Initiatives, Kitware
